File: sdk/java-sdk/src/main/java/com/evernym/verity/sdk/protocols/Handlers.java
package com.evernym.verity.sdk.protocols;

import java.util.ArrayList;
import java.util.concurrent.ExecutionException;

import com.evernym.verity.sdk.utils.MessagePackaging;
import com.evernym.verity.sdk.utils.VerityConfig;

import org.hyperledger.indy.sdk.IndyException;
import org.json.JSONObject;

/**
 * Stores an array of message handlers that are used when receiving an inbound message
 */
public class Handlers {
    static ArrayList<MessageHandler> messageHandlers = new ArrayList<MessageHandler>();

    /**
     * Adds a MessageHandler for a message type to the list if current message handlers
     * @param messageType the type of message to be handled
     * @param messageHandler the handler function itself
     */
    public static void addHandler(String messageType, MessageHandler.Handler messageHandler) {
        messageHandlers.add(new MessageHandler(messageType, messageHandler));
    }

    /**
     * Adds a MessageHandler for a message type and status to the list if current message handlers
     * @param messageType the type of message to be handled
     * @param messageStatus the status of the message to be handled
     * @param messageHandler the handler function itself
     */
    public static void addHandler(String messageType, Integer messageStatus, MessageHandler.Handler messageHandler) {
        messageHandlers.add(new MessageHandler(messageType, messageStatus, messageHandler));
    }

    /**
     * Calls all of the handlers that support handling of this particular message type and message status
     * @param verityConfig an instance of VerityConfig configured with the results of the provision_sdk.py script
     * @param rawMessage the raw bytes received from Verity
     * @throws InterruptedException when there are issues with encryption and decryption
     * @throws ExecutionException when there are issues with encryption and decryption
     * @throws IndyException when there are issues with encryption and decryption
     */
    public static void handleMessage(VerityConfig verityConfig, byte[] rawMessage) throws InterruptedException, ExecutionException, IndyException {
        JSONObject message = MessagePackaging.unpackMessageFromVerity(verityConfig, rawMessage);
        System.out.println("New message from verity: " + message.toString());
        for(MessageHandler messageHandler: messageHandlers) {
            if(messageHandler.handles(message)) {
                messageHandler.handle(message);
            }
        }
    }
    
}

File: samples/sdk/dotnet-example-app-tmp/VeritySDK/Handlers/Handlers.cs
using System.Collections.Generic;
using System.Json;
using VeritySDK.Protocols;
using VeritySDK.Utils;

namespace VeritySDK.Handler
{
    /// <summary>
    /// Stores an array of message handlers that are used when receiving an inbound message
    /// </summary>
    public class Handlers
    {
        /// <summary>
        /// Delegate for handles the given JSON object message
        /// </summary>
        /// <param name="message">the message to be handled</param>
        public delegate void DefaultHandler(JsonObject message);


        private DefaultHandler defaultHandler;
        private List<MessageHandler> messageHandlers = new List<MessageHandler>();

        /// <summary>
        /// Adds a handler for all message types not handled by other message handlers
        /// </summary>
        /// <param name="messageHandler">the function that will be called</param>
        public void addDefaultHandler(DefaultHandler messageHandler)
        {
            defaultHandler = messageHandler;
        }

        /// <summary>
        /// Adds a MessageHandler for a message type to the list if current message handlers
        /// </summary>
        /// <param name="messageFamily">the family of the message to be handled</param>
        /// <param name="messageHandler">the handler function itself</param>
        public void addHandler(MessageFamily messageFamily, MessageHandler.Handler messageHandler)
        {
            messageHandlers.Insert(0, new MessageHandler(messageFamily, messageHandler));
        }

        /// <summary>
        /// Calls all of the handlers that support handling of this particular message type and message status
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <param name="rawMessage">the raw bytes received from Verity</param>
        public void handleMessage(Context context, byte[] rawMessage)
        {
            JsonObject message = Util.unpackMessage(context, rawMessage);
            foreach (MessageHandler messageHandler in messageHandlers)
            {
                if (messageHandler.handles(message))
                {
                    messageHandler.handle(message);
                    return;
                }
            }

            // call default if another handler is not called
            if (defaultHandler != null)
                defaultHandler(message);
        }
    }
}
